src/conversions/std/cstring.rs

```diff
@@ -0,0 +1,186 @@
+use crate::types::PyString;
+use crate::{Borrowed, Bound, FromPyObject, IntoPyObject, PyAny, PyErr, Python};
+use std::borrow::Cow;
+use std::ffi::{CStr, CString};
+use std::str::Utf8Error;
+#[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+use {
+    crate::{exceptions::PyValueError, ffi},
+    std::slice,
+};
+
+impl<'py> IntoPyObject<'py> for &CStr {
+    type Target = PyString;
+    type Output = Bound<'py, Self::Target>;
+    type Error = Utf8Error;
+
+    #[inline]
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        self.to_str()?.into_pyobject(py).map_err(|err| match err {})
+    }
+}
+
+impl<'py> IntoPyObject<'py> for CString {
+    type Target = PyString;
+    type Output = Bound<'py, Self::Target>;
+    type Error = Utf8Error;
+
+    #[inline]
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        (&*self).into_pyobject(py)
+    }
+}
+
+impl<'py> IntoPyObject<'py> for &CString {
+    type Target = PyString;
+    type Output = Bound<'py, Self::Target>;
+    type Error = Utf8Error;
+
+    #[inline]
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        (&**self).into_pyobject(py)
+    }
+}
+
+impl<'py> IntoPyObject<'py> for Cow<'_, CStr> {
+    type Target = PyString;
+    type Output = Bound<'py, Self::Target>;
+    type Error = Utf8Error;
+
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        (*self).into_pyobject(py)
+    }
+}
+
+impl<'py> IntoPyObject<'py> for &Cow<'_, CStr> {
+    type Target = PyString;
+    type Output = Bound<'py, Self::Target>;
+    type Error = Utf8Error;
+
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        (&**self).into_pyobject(py)
+    }
+}
+
+#[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+impl<'a> FromPyObject<'a, '_> for &'a CStr {
+    type Error = PyErr;
+
+    fn extract(obj: Borrowed<'a, '_, PyAny>) -> Result<Self, Self::Error> {
+        let obj = obj.cast::<PyString>()?;
+        let mut size = 0;
+        let ptr = unsafe { ffi::PyUnicode_AsUTF8AndSize(obj.as_ptr(), &mut size) };
+
+        if ptr.is_null() {
+            return Err(PyErr::fetch(obj.py()));
+        }
+
+        // SAFETY: PyUnicode_AsUTF8AndSize always returns a NUL-terminated string
+        let slice = unsafe { slice::from_raw_parts(ptr.cast(), size as usize + 1) };
```

Comment:
Maybe spell out a bit more why the `+1` is safe? I'm thinking that the `size` is the length of the string not including the `NUL`, but we can safely read that extra byte? But it would be better to be explicit about that.